### PR TITLE
Midi WIP

### PR DESCRIPTION
**Pull request by hhoegelo**
- decouple outgoing external midi from audio thread
- add lsb for hardware soucres and velocity
- use CC 20..25 for hardware sources MSB and 52..57 LSB

### CODE REVIEW

**Comment by hhoegelo** [comment]
@MatthiasSeeber Changing the range of CC was requested by @nonlinear-labs

File: projects/epc/audio-engine/src/synth/c15-audio-engine/dsp_host_dual.cpp [hunk 1]
```diff
@@ -1447,14 +1490,77 @@ float dsp_host_dual::scale(const Scale_Aspect _scl, float _value)
...
+
+    case 4:  // Pitch
+    {
+      uint8_t statusByte = static_cast<uint8_t>(0xE0);
+      uint16_t v = static_cast<uint16_t>(value * (1 << 13));
+      uint8_t valByte1 = static_cast<uint8_t>(v & 0x7F);
+      uint8_t valByte2 = static_cast<uint8_t>((v >> 7) & 0x7F);
+      out({ statusByte, valByte1, valByte2 });
+      break;
+    }
```
Comment (MatthiasSeeber) [inline projects/epc/audio-engine/src/synth/c15-audio-engine/dsp_host_dual.cpp:1544]: The "Pitch" is not a Hardware Source. This `update_hw` function triggers the whole Modulation Process. It should not be triggered by a Note On/Off message. I'd suggest to process the pitch in a separate function.

File: projects/epc/audio-engine/src/synth/c15-audio-engine/dsp_host_dual.cpp [hunk 2]
```diff
@@ -527,48 +543,75 @@ void dsp_host_dual::onMidiMessage(const uint32_t _status, const uint32_t _data0,
       onTcdMessage(0xED, 0, _data0);  // keyPos
       if(arg != 0)
       {
-        onTcdMessage(0xEE, arg >> 7, arg & 127);  // keyDown
+        onTcdMessage(0xEE, arg >> 7, std::exchange(m_velocityLSB, 0));  // keyDown
```
Comment (MatthiasSeeber) [inline projects/epc/audio-engine/src/synth/c15-audio-engine/dsp_host_dual.cpp:546]: I have trouble understanding this. Why isn't this needed in `[538] onTcdMessage(0xEF, arg >> 7, arg & 127);  // keyUp`?

File: projects/epc/audio-engine/src/synth/c15-audio-engine/dsp_host_dual.cpp [hunk 3]
```diff
@@ -474,6 +485,11 @@ void dsp_host_dual::onTcdMessage(const uint32_t _status, const uint32_t _data0,
           auto vel = static_cast<float>(arg) * m_norm_vel;
           keyDown(vel);
 
+          uint8_t highResolutionVelocityStatusByte = static_cast<uint8_t>(0xB0);
+          uint16_t fullResolutionValue = vel * (1 << 13);
```
Comment (MatthiasSeeber) [inline projects/epc/audio-engine/src/synth/c15-audio-engine/dsp_host_dual.cpp:489]: Will the compiler treat `(1 << 13)` as a constexpr? (I so far preferred initializing a constant variable for this, avoding such shifts within a function)

**Review by MatthiasSeeber — COMMENTED** [review]
Let's talk about this, and try to find the reason for the shifted pitches..

File: projects/epc/audio-engine/src/synth/c15-audio-engine/dsp_host_dual.cpp [hunk 4]
```diff
@@ -474,6 +485,11 @@ void dsp_host_dual::onTcdMessage(const uint32_t _status, const uint32_t _data0,
...
+          uint8_t lsbValByte = static_cast<uint8_t>(fullResolutionValue & 0x7F);
+          out({ highResolutionVelocityStatusByte, 88, lsbValByte });
+
           uint8_t statusByte = static_cast<uint8_t>(0x90);
           uint8_t keyByte = static_cast<uint8_t>(m_key_pos) & 0x7F;
```
Comment (MatthiasSeeber) [inline projects/epc/audio-engine/src/synth/c15-audio-engine/dsp_host_dual.cpp:494]: I think, here arises the problem of the shifted pitch. In `[472] m_key_pos = arg - C15::Config::key_from;` is shifted to transform a "Note Pitch" (36..96) into a "Key Position" (0..60). Futher using the `m_key_pos` like a "Note Pitch" requires adding `C15::Config::key_from` again.

File: projects/epc/audio-engine/src/synth/c15-audio-engine/dsp_host_dual.cpp (hunk 3 above)
Comment (hhoegelo) [reply projects/epc/audio-engine/src/synth/c15-audio-engine/dsp_host_dual.cpp:489]: Yes, the expression will be replaced with a const literal value by the compiler.

File: projects/epc/audio-engine/src/synth/c15-audio-engine/dsp_host_dual.cpp (hunk 4 above)
Comment (hhoegelo) [reply projects/epc/audio-engine/src/synth/c15-audio-engine/dsp_host_dual.cpp:494]: Okay thanks - will fix it.

File: projects/epc/audio-engine/src/synth/c15-audio-engine/dsp_host_dual.cpp (hunk 2 above)
Comment (hhoegelo) [reply projects/epc/audio-engine/src/synth/c15-audio-engine/dsp_host_dual.cpp:546]: You are right - I will add it for note off.

File: projects/epc/audio-engine/src/synth/c15-audio-engine/dsp_host_dual.cpp (hunk 1 above)
Comment (hhoegelo) [reply projects/epc/audio-engine/src/synth/c15-audio-engine/dsp_host_dual.cpp:1544]: It is just a wrong comment - it should read "Bender".
Would this fix the problem or do I still miss something?

**Review by schlenkibus — APPROVED** [review]

**Review by MatthiasSeeber — APPROVED** [review]

**Pull request merged**